stdlib/universe/group.go

```diff
@@ -135,84 +139,231 @@ func createGroupTransformation(id execute.DatasetID, mode execute.AccumulationMo
 	if !ok {
 		return nil, nil, errors.Newf(codes.Internal, "invalid spec type %T", spec)
 	}
-	cache := execute.NewTableBuilderCache(a.Allocator())
-	d := execute.NewDataset(id, mode, cache)
-	t := NewGroupTransformation(d, cache, s)
+	t, d := NewGroupTransformation(s, id, a.Allocator())
 	return t, d, nil
 }
 
 type groupTransformation struct {
 	d     execute.Dataset
-	cache execute.TableBuilderCache
+	cache tableutil.BuilderCache
+	mem   *memory.Allocator
 
 	mode flux.GroupMode
 	keys []string
 }
 
-func NewGroupTransformation(d execute.Dataset, cache execute.TableBuilderCache, spec *GroupProcedureSpec) *groupTransformation {
+func NewGroupTransformation(spec *GroupProcedureSpec, id execute.DatasetID, mem *memory.Allocator) (execute.Transformation, execute.Dataset) {
```

Comment:
Filter was also changed for this. It is something I have been experimenting with. It is mostly because I didn't see much of a point in separating out these parameters. I can't think of a circumstance where the passed in `execute.Dataset` is actually pluggable and it would be a pretty bad error if the dataset was not constructed with the same cache that is passed in. I am not sure if the difference between a transformation and a dataset is really very important at the moment, but I also didn't want to rewrite the engine to improve performance when I didn't know what the final design should be.

When I was changing filter, it also seemed to me like the dataset was very implementation dependent. And if the implementation was dependent on the transformation, I couldn't find a way to make the dataset interface work in a way that functioned with all of the different types of transformations. So I thought it would be best if the transformation got to decide what dataset it output to.

I do think this probably needs to be improved. I don't have a strong rationale for why the dataset is constructed in this way but I also don't have a very strong rationale for the old design either. I had been thinking it might help to move all of these methods into an internal package since this method is not needed outside of the flux repository.

Reply:
Can we give the transformation a dataset getter so it doesn't have to be returned in the function?

Reply:
Maybe. I'm not really sure and I'm hesitant to touch that section of code any more than I already have until we decide we're going to commit to a refactor of the engine.